src/async/control-flow.md

```diff
@@ -0,0 +1,200 @@
+# Futures Control Flow
+
+Futures can be combined together to produce concurrent compute flow graphs.
+We will demonstrate here a non-exhaustive array of common operations:
+
+## Run a group of futures concurrently until they all resolve: `join_all`
+
+### Equivalents:
+
+- JS: `Promise.all`
+- Python: `asyncio.gather`
+
+```rust,editable,compile_fail
+use anyhow::Result;
+use futures::future;
+use reqwest;
+use std::collections::HashMap;
+
+async fn size_of_page(url: &str) -> Result<usize> {
+    let resp = reqwest::get(url).await?;
+    Ok(resp.text().await?.len())
+}
+
+#[tokio::main]
+async fn main() {
+    let urls: [&str; 4] = [
+        "https://google.com",
+        "https://httpbin.org/ip",
+        "https://play.rust-lang.org/",
+        "BAD_URL",
+    ];
+    let futures_iter = urls.into_iter().map(size_of_page);
+    let results = future::join_all(futures_iter).await;
+    let page_sizes_dict: HashMap<&str, Result<usize>> =
+        urls.into_iter().zip(results.into_iter()).collect();
+    println!("{:?}", page_sizes_dict);
+}
+```
+
+## Run multiple futures concurrently until the first one resolves: `select`
+
+### Equivalents:
+
+- JS: `Promise.race`
+- Python: `asyncio.new_event_loop().run_until_complete(asyncio.wait(task_set, return_when=asyncio.FIRST_COMPLETED))`
+
+```rust,editable,compile_fail
+use anyhow::Result;
+use tokio::sync::mpsc::{self, Receiver, Sender};
+use tokio::time::{sleep, Duration};
+
+#[derive(Debug, PartialEq)]
+enum Animal {
+    Cat { name: String },
+    Dog { name: String },
+}
+
+async fn first_animal_to_finish_race(
+    mut cat_rcv: Receiver<String>,
+    mut dog_rcv: Receiver<String>,
+) -> Option<Animal> {
+    tokio::select! {
+        cat_name = cat_rcv.recv() => Some(Animal::Cat { name: cat_name? }),
+        dog_name = dog_rcv.recv() => Some(Animal::Dog { name: dog_name? })
+    }
+}
+
+async fn simulate_animals_race(
+    sender: Sender<String>,
+    duration: Duration,
+    name: String,
+) -> Result<()> {
+    sleep(duration).await;
+    sender.send(name).await?;
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() {
+    let (cat_sender, cat_receiver) = mpsc::channel(32);
+    let (dog_sender, dog_receiver) = mpsc::channel(32);
+    tokio::spawn(simulate_animals_race(
+        cat_sender,
+        Duration::from_secs(10),
+        String::from("Felix"),
+    ));
+    tokio::spawn(simulate_animals_race(
+        dog_sender,
+        Duration::from_secs(5),
+        String::from("Rex"),
+    ));
+
+    let winner = first_animal_to_finish_race(cat_receiver, dog_receiver)
+        .await
+        .expect("Failed to receive winner");
+
+    println!("The winner of the race is {:?}", winner);
+
+    assert_eq!(
+        winner,
+        Animal::Dog {
+            name: String::from("Rex")
+        }
+    );
+}
+```
```

Comment:
This example seems a bit lengthy. Could this start with a shorter, simpler example with ideas in the speaker notes to improve it?